src/arbitrary/_std/string.rs

```diff
@@ -0,0 +1,247 @@
+//-
+// Copyright 2017, 2018 The proptest developers
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Arbitrary implementations for `std::string`.
+
+use std::string::{String, FromUtf8Error, FromUtf16Error};
+use std::iter;
+use std::slice;
+use std::rc::Rc;
+use std::sync::Arc;
+
+use strategy::*;
+use strategy::statics::static_map;
+use collection::SizeRange;
+use arbitrary::*;
+
+/// Wraps the regex that forms the `Strategy` for `String` so that a sensible
+/// `Default` can be given. The default is a string of non-control characters.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct StringParam(&'static str);
+
+impl From<StringParam> for &'static str {
+    fn from(x: StringParam) -> Self { x.0 }
+}
+
+impl From<&'static str> for StringParam {
+    fn from(x: &'static str) -> Self { StringParam(x) }
+}
+
+impl Default for StringParam {
+    fn default() -> Self {
+        StringParam("\\PC*")
+    }
+}
+
+impl Arbitrary for String {
+    valuetree!();
+    type Parameters = StringParam;
+    type Strategy = &'static str;
+
+    /// ## Safety
+    ///
+    /// This implementation panics if the input is not a valid regex proptest
+    /// can handle.
+    fn arbitrary_with(args: Self::Parameters) -> Self::Strategy {
+        args.into()
+    }
+}
+
+macro_rules! dst_wrapped {
+    ($($w: ident),*) => {
+        $(arbitrary!($w<str>, MapInto<StrategyFor<String>, Self>, StringParam;
+            a => any_with::<String>(a).prop_map_into()
+        );)*
+    };
+}
+
+dst_wrapped!(Box, Rc, Arc);
+
+lazy_just!(FromUtf16Error, || String::from_utf16(&[0xD800]).unwrap_err());
+
+// This is a void-like type, it needs to be handled by the user of
+// the type by simply never constructing the variant in an enum or for
+// structs by inductively not generating the struct.
+// The same applies to ! and Infallible.
+// generator!(ParseError, || panic!());
+
+arbitrary!(FromUtf8Error, SFnPtrMap<BoxedStrategy<Vec<u8>>, Self>;
+    static_map(not_utf8_bytes(true), |bs| String::from_utf8(bs).unwrap_err())
+);
+
+// This could be faster.. The main cause seems to be generation of
+// Vec<char>. any::<u8>() instead of any::<u16>() speeds it up considerably.
+pub(crate) fn not_utf8_bytes(allow_null: bool) -> BoxedStrategy<Vec<u8>> {
```

Comment:
I truly hope I didn't mess this one up as it probably took 1/10 of the time =)
I should probably document this better.. `EL` stands for `error_length` and is meant to result in different values for [`.error_len()`](https://doc.rust-lang.org/nightly/std/str/struct.Utf8Error.html#method.error_len).
What I did was reverse the logic in `run_utf8_validation` so that the strategy produces arbitrary non-UTF8 bytes by construction focusing on as uniform a distribution as possible for the values produced by `error_len` of `Utf8Error`.

Hmm.. is `collection::vec(1u8.., size)` guaranteed to always generate invalid UTF-8?

Reply:
I see, I misunderstood where this was being used.

> is `collection::vec(1u8.., size)` guaranteed to always generate invalid UTF-8?

No. I thought this was all trying to produce strings with possibly invalid UTF-8, which is why I was so puzzled as to why there was so much logic here.

I'll take me a bit to fully understand this bit, so I'll get back to this thread later

Reply:
I updated the definition of `not_utf8_bytes` and got rid of the `prop_flat_map` there =)
I also tried to document it better.